crmprtd/align.py

```diff
@@ -8,12 +8,237 @@
+                                             obs_tuple.lon,
+                                             float(lat),
+                                             float(lon))
+
+                    # best case, we find exact match
+                    if dist == 0:
+                        log.debug('Exact match found')
+                        hid = id
+                        break
+
+                    # find closest
+                    else:
+                        if dist < closest_pos:
+                            hid = id
+                            closest_pos = dist
+
+            if not hid:
+                log.debug('Observation contains lat/lon but does not match '
+                          'any entries in db')
+                hid = create_station_and_history_entry(sesh, obs_tuple)
+        else:
+            log.debug('No lat/lon, checking for \"active\" stations')
+            for id in q.all():
+                q = sesh.query(History.id).filter(
+                    and_(History.id == id,
+                         History.sdate is not None,
+                         History.edate is None))
+
+                if q.count() > 0:
+                    log.debug('Matched hid using sdate/edate')
+                    hid, = q.first()
+                    break
+
+        if not hid:
+            log.warning('Unable to match station',
+                        extra={'station_id': obs_tuple.station_id,
+                               'network_name': obs_tuple.network_name})
+            return
+
+    # thing
+    log.debug('Check time')
+    if obs_tuple.time is None:
+        log.error('Observation cannot be used without time')
+        return
+    log.debug('Observation has time')
+    time = obs_tuple.time
+
+    log.debug('Check data')
+    if obs_tuple.val is None:
+        log.error('Observation cannot be used without value')
+        return
+    log.debug('Observation has value')
+    datum = obs_tuple.val
+
+    log.debug('Check if variable name exists in database')
+    q = sesh.query(Variable.id).join(Network).filter(and_(
+        Network.name == obs_tuple.network_name,
+        Variable.name == obs_tuple.variable_name))
+
+    if q.count() < 1:
+        log.warning('No matching variable found', extra={
+                    'vars_id': obs_tuple.variable_name})
+        return
+
+    vars_id, = q.first()
+    log.debug('Observation variable matches', extra={'vars_id': vars_id})
+
+    log.debug('Check unit')
+    if obs_tuple.unit is None:
+        log.debug('No unit given, checking if table contains unit')
+        unit, = sesh.query(Variable.unit).join(Network).filter(and_(
+            Network.name == obs_tuple.network_name,
+            Variable.name == obs_tuple.variable_name)).first()
+
+        if not unit:
+            log.warning('Table contains no unit for variable',
+                        extra={'var_name': obs_tuple.variable_name})
+            return
```

Comment:
Going back to the empty return, this function is a generator since you are yielding things at the bottom. What happens if you yield a few things, return None, then try to yield a few more things? You'll never get to the things after the return (and in some versions of Python, mixing returns and yields in a single function is a syntax error).

```python
$ python
Python 3.6.5 (default, May  3 2018, 10:08:28) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo():
...     yield 1
...     yield 2
...     yield 3
...     return
...     yield 4
...     yield 5
... 
>>> for i in foo():
...     print(i)
... 
1
2
3
```

So you want to better define what happens in an error condition here. You *probably* want to log.error and then `continue` to the next iteration.

Reply:
**Update:** I forgot that the argument to this function is just a single Obs tuple and the result is a single Obs object. That being the case, it's probably appropriate to return None (i.e. an empty return statement), but then you'll want to `return` the Obs at the bottom of the function, rather than `yield`ing it. There's not really much point to a `yield` of a single object, since you're not going to iterate over a single thing.